Sources/Plasma/NucleusLib/pnAsyncCoreExe/pnAceSocket.cpp

```diff
@@ -0,0 +1,553 @@
+    std::array<AsyncThreadRef, kMaxWorkerThreads> fThreadHandles;
+
+    AsyncIoPool() : fWorkGuard(fContext.get_executor())
+    {
+        // calculate number of IO worker threads to create
+        unsigned int threadCount = std::max(std::thread::hardware_concurrency() * 2, 2U);
+        if (threadCount > kMaxWorkerThreads) {
+            threadCount = kMaxWorkerThreads;
+            LogMsg(kLogError, "kMaxWorkerThreads too small!");
+        }
+
+        // create IO worker threads
+        for (unsigned int thread = 0; thread < threadCount; thread++) {
+            fThreadHandles[thread] = AsyncThreadCreate([this] {
+                // This can be run concurrently from several threads
+                fContext.run();
+            });
+        }
+    }
+
+    void Destroy(unsigned exitThreadWaitMs)
+    {
+        fWorkGuard.reset();
+
+        for (AsyncThreadRef& thread : fThreadHandles) {
+            if (thread.joinable())
+                AsyncThreadTimedJoin(thread, exitThreadWaitMs);
+        }
+
+        // Ensure the event loop exits without processing any more tasks,
+        // in case the threads take more than exitThreadWaitMs to finish
+        fContext.stop();
+    }
+};
+
+struct ConnectOperation
+{
+    tcp::socket             fSock;
+    plNetAddress            fRemoteAddr;
+    AsyncCancelId           fCancelId;
+    FAsyncNotifySocketProc  fNotifyProc;
+    void*                   fParam;
+    std::vector<uint8_t>    fConnectBuffer;
+    unsigned int            fConnectionType;
+
+    ConnectOperation(asio::io_context& context)
+        : fSock(context), fCancelId(), fParam(), fConnectionType()
+    { }
+};
+
+struct WriteOperation
+{
+    unsigned int            fAllocSize;
+    AsyncNotifySocketWrite  fNotify;
+    unsigned                queueTimeMs;
+
+    WriteOperation() : fAllocSize() { }
+
+    asio::const_buffer AsBuffer() const
+    {
+        return asio::buffer(fNotify.buffer + fNotify.bytesCommitted, fNotify.bytes - fNotify.bytesCommitted);
+    }
+};
+
+struct AsyncSocketStruct
+{
+    std::recursive_mutex        fCritsect;
+    tcp::socket                 fSock;
+    FAsyncNotifySocketProc      fNotifyProc;
+    void*                       fParam;
+    unsigned int                fConnectionType;
+    void*                       fUserState;
+    uint8_t                     fBuffer[kAsyncSocketBufferSize];
+    size_t                      fBytesLeft;
+    std::list<WriteOperation *> fWriteOps;
+    unsigned                    initTimeMs;
+    unsigned                    closeTimeMs;
+
+    AsyncSocketStruct(ConnectOperation* op)
+        : fSock(std::move(op->fSock)), fNotifyProc(op->fNotifyProc),
+          fParam(op->fParam), fConnectionType(op->fConnectionType),
+          fUserState(), fBuffer(), fBytesLeft(), initTimeMs(), closeTimeMs()
+    { }
+};
+
+static AsyncIoPool*                 s_ioPool;
+
+static std::recursive_mutex         s_connectCrit;
+static std::list<ConnectOperation*> s_connectList;

```

Comment:
It's not clear to me this change should be made. I don't know if I have the expertise in C++ or ASIO to make the call. This struct contains values that will be made invalid in the source struct the struct is copied. Strict passing by reference could sidestep this issue, but passing by pointer could be a way to make sure the struct is not unintentionally copied.

Reply:
You can delete the copy constructor to make sure that the struct isn't copied accidentally. In the struct's decalration:

```C++
struct ConnectOperation
{
    ConnectOperation(const ConnectOperation&) = delete;
};
```

Reply:
Oh right, I'll take another look then.

Reply:
Ok, took another look. In pnAceSocket.cpp:350 there is this:
```
        // This operation is no longer a candidate for cleanup or cancellation
        // by the time we get a callback from async_connect
        {
            hsLockGuard(s_connectCrit);
            auto iter = std::find(s_connectList.cbegin(), s_connectList.cend(), op);
            if (iter != s_connectList.cend())
                s_connectList.erase(iter);
        }
```

A comparison operator could be written. But I'm starting to think maybe this is just best left as a pointer... Unless you have any other tricks! :)

Reply:
That's a good point in that I missed something in my code review. The snippet that you've referenced could be done in constant time instead of by re-looking up the connection in the list by having `std::list` return an iterator to the newly created item. `std::list` iterators are guaranteed to be stable (read: valid) until the item is erased from the container. This will also remove extra pointer indirection, potentially improving cache utilization. For example, I would rewrite https://github.com/H-uru/Plasma/pull/1205/files/21db57551b4b0cc60cbe1c2d09b1f37884d0aca7#diff-ac6773cf2946916e561937a5d6a0b1fafd0245b5412722d5b1a5933531e51542R311-R378 to look like:

```c++
void AsyncSocketConnect(AsyncCancelId* cancelId, const plNetAddress& netAddr,
                        FAsyncNotifySocketProc notifyProc, void* param,
                        const void* sendData, unsigned sendBytes)
{
    ASSERT(notifyProc);
    ASSERT(s_ioPool);

    PerfAddCounter(kAsyncPerfSocketConnAttemptsOutCurr, 1);
    PerfAddCounter(kAsyncPerfSocketConnAttemptsOutTotal, 1);

    // Locking the critical section here will not deadlock because `asio::basic_socket::async_connect`
    // will return immediately. Therefore, the lock in the callback cannot deadlock us.
    hsLockGuard(s_connectCrit);
    auto op = s_connectList.emplace(s_connectList.end(), s_ioPool->fContext);
    op->fRemoteAddr = netAddr;
    op->fNotifyProc = std::move(notifyProc);
    op->fParam = param;
    if (sendBytes) {
        auto sendBuffer = reinterpret_cast<const uint8_t*>(sendData);
        op->fConnectBuffer.assign(sendBuffer, sendBuffer + sendBytes);
        op->fConnectionType = sendBuffer[0];
    } else {
        op->fConnectionType = kConnTypeNil;
    }

    if (cancelId) {
        // get cancel id; we can avoid checking for zero by always using an odd number
        ASSERT(s_nextConnectCancelId & 1);
        s_nextConnectCancelId += 2;

        *cancelId = op->fCancelId = (AsyncCancelId)s_nextConnectCancelId;
    }

    tcp::endpoint remoteAddr(asio::ip::address_v4(netAddr.GetHostBytes()),
                             netAddr.GetPort());
    op->fSock.async_connect(remoteAddr, [op](const asio::error_code& connError) {
        // This operation is no longer a candidate for cleanup or cancellation
        // by the time we get a callback from async_connect
        bool success;
        if (connError) {
            LogMsg(kLogError, "Failed to connect: {}", connError.message());
            success = false;
        } else {
            success = SocketInitConnect(op);
        }

        if (!success) {
            AsyncNotifySocketConnect failed;
            failed.param      = op->fParam;
            failed.connType   = op->fConnectBuffer[0];
            failed.remoteAddr = op->fRemoteAddr;
            failed.localAddr.Clear();
            op->fNotifyProc(nullptr, kNotifySocketConnectFailed, &failed, nullptr);
        }

        {
            hsLockGuard(s_connectCrit);
            s_connectList.erase(op);
        }

        PerfSubCounter(kAsyncPerfSocketConnAttemptsOutCurr, 1);
    });
}
```

I also added a comment about why what appeared to me to be a potential deadlock is not actually a deadlock :)

Reply:
Let me take a look at implementing your snippet and the other required changes. Did not know that was a feature of C++ iterators!

Reply:
Each type of container has slightly different iterator rules. It's generally best to assume that iterators are not stable, especially for containers like std::vector. The list iterator is IIUC basically a fancy reference to a node in the linked list, so this is fine :)